File: example-clients/python/sse_client.py
#!/usr/bin/env python3
"""Consume a WorldSignal subscription as a Server-Sent Events stream.

Env:
  WS_API_BASE     base URL (default http://localhost:4800)
  WS_API_KEY      API key with the signals:read scope (required)
  WS_SUBSCRIPTION subscription id (default demo-stream)
  WS_SINCE        resume cursor / seq (default: 0 = from the start)
  WS_MAX          exit after N events (default: 0 = run forever)

Uses only the standard library. Prefer the Authorization header (below) over
the ?api_key= query param, which browsers need but proxies may log.
"""
import json
import os
import sys
import urllib.request

BASE = os.environ.get("WS_API_BASE", "http://localhost:4800")
KEY = os.environ.get("WS_API_KEY", "")
SUB = os.environ.get("WS_SUBSCRIPTION", "demo-stream")
SINCE = os.environ.get("WS_SINCE", "0")
MAX = int(os.environ.get("WS_MAX", "0"))


def main() -> int:
    if not KEY:
        print("WS_API_KEY is required", file=sys.stderr)
        return 2
    url = f"{BASE}/v1/stream/sse?subscription={SUB}&since={SINCE}"
    req = urllib.request.Request(url, headers={"Authorization": f"Bearer {KEY}"})
    print(f"[sse] connecting to {url}", file=sys.stderr)

    seen = 0
    with urllib.request.urlopen(req) as resp:
        for raw in resp:
            line = raw.decode("utf-8", "replace").rstrip("\n")
            if not line.startswith("data:"):
                continue  # skip id:/event:/comment lines
            event = json.loads(line[len("data:"):].strip())
            d = event.get("data", {})
            print(f"[sse] {d.get('severity','?'):8} {d.get('country') or '--'}  {d.get('title','')}")
            seen += 1
            if MAX and seen >= MAX:
                print(f"[sse] received {seen} event(s), exiting", file=sys.stderr)
                return 0
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: example-clients/php/sse_client.php
<?php
// Consume a WorldSignal subscription as Server-Sent Events — no extensions.
// Opens the stream with an Authorization header and parses `data:` lines.
// Env: WS_API_BASE, WS_API_KEY, WS_SUBSCRIPTION, WS_SINCE, WS_MAX.

$base = getenv('WS_API_BASE') ?: 'http://localhost:4800';
$key = getenv('WS_API_KEY') ?: '';
$sub = getenv('WS_SUBSCRIPTION') ?: 'demo-stream';
$since = getenv('WS_SINCE') ?: '0';
$max = (int)(getenv('WS_MAX') ?: '0');

if ($key === '') { fwrite(STDERR, "WS_API_KEY is required\n"); exit(2); }

$url = "$base/v1/stream/sse?subscription=$sub&since=$since";
fwrite(STDERR, "[sse] connecting to $url\n");
$ctx = stream_context_create(['http' => [
    'method' => 'GET',
    'header' => "Authorization: Bearer $key\r\n",
    'timeout' => 30,
]]);
$fp = fopen($url, 'r', false, $ctx);
if ($fp === false) { fwrite(STDERR, "[sse] connection failed\n"); exit(1); }

$seen = 0;
while (($line = fgets($fp)) !== false) {
    $line = rtrim($line, "\r\n");
    if (strncmp($line, 'data:', 5) !== 0) { continue; } // skip id:/event:/comments
    $ev = json_decode(trim(substr($line, 5)), true);
    $d = $ev['data'] ?? [];
    printf("[sse] %-8s %s  %s\n", $d['severity'] ?? '?', $d['country'] ?? '--', $d['title'] ?? '');
    if ($max > 0 && ++$seen >= $max) {
        fwrite(STDERR, "[sse] received $seen event(s), exiting\n");
        fclose($fp);
        exit(0);
    }
}
fclose($fp);
